Concurrent write to websocket connection

The following error may occur when sending multiple messages concurrently. As mentioned in the doc https://godoc.org/github.com/gorilla/websocket#hdr-Concurrency we need to add mutex to make sure we don't call writeMessage concurrently.

```
panic: concurrent write to websocket connection

goroutine 59 [running]:
github.com/gorilla/websocket.(*messageWriter).flushFrame(0xc0001419e0, 0xc000067d01, 0x0, 0x0, 0x0, 0x64, 0x3)
	/Users/skysniper/go/pkg/mod/github.com/gorilla/websocket@v1.4.0/conn.go:591 +0x77a
github.com/gorilla/websocket.(*messageWriter).Close(0xc0001419e0, 0x64, 0xc000067d70)
	/Users/skysniper/go/pkg/mod/github.com/gorilla/websocket@v1.4.0/conn.go:709 +0x56
github.com/gorilla/websocket.(*Conn).prepWrite(0xc0002242c0, 0x2, 0x70, 0x1629b60)
	/Users/skysniper/go/pkg/mod/github.com/gorilla/websocket@v1.4.0/conn.go:459 +0x20d
github.com/gorilla/websocket.(*Conn).NextWriter(0xc0002242c0, 0x2, 0x68, 0xc00000f400, 0xc00007e4d0, 0x68)
	/Users/skysniper/go/pkg/mod/github.com/gorilla/websocket@v1.4.0/conn.go:494 +0x39
github.com/gorilla/websocket.(*Conn).WriteMessage(0xc0002242c0, 0x2, 0xc00007e4d0, 0x68, 0x68, 0x0, 0x0)
	/Users/skysniper/go/pkg/mod/github.com/gorilla/websocket@v1.4.0/conn.go:755 +0x73
github.com/nknorg/nkn-sdk-go.(*Client).sendReceipt(0xc0001bc000, 0xc0000d31c0, 0x20, 0x20, 0xc000096420, 0x0)
	/Users/skysniper/go/pkg/mod/github.com/nknorg/nkn-sdk-go@v0.0.0-20190710235647-86437d8d12e5/client.go:281 +0x407
github.com/nknorg/nkn-sdk-go.(*Client).connect.func2.2.1(0xc0001bc000, 0xc0002ee000)
	/Users/skysniper/go/pkg/mod/github.com/nknorg/nkn-sdk-go@v0.0.0-20190710235647-86437d8d12e5/client.go:200 +0x4f
created by github.com/nknorg/nkn-sdk-go.(*Client).connect.func2.2
	/Users/skysniper/go/pkg/mod/github.com/nknorg/nkn-sdk-go@v0.0.0-20190710235647-86437d8d12e5/client.go:199 +0x2e5
```

--- a/client.go
+++ b/client.go
@@ -7,6 +7,7 @@ import (
 	"encoding/json"
 	"log"
 	"net/url"
+	"sync"
 	"time"
 
 	"github.com/gogo/protobuf/proto"
@@ -34,7 +35,6 @@ type Client struct {
 	Address   string
 	addressId []byte
 	urlString string
-	conn      *websocket.Conn
 	closed    bool
 	OnConnect chan struct{}
 	OnMessage chan *pb.InboundMessage
@@ -42,6 +42,9 @@ type Client struct {
 
 	nodeInfo          *NodeInfo
 	sigChainBlockHash string
+
+	sync.Mutex
+	conn *websocket.Conn
 }
 
 type NodeInfo struct {
@@ -232,8 +235,8 @@ func NewClient(account *vault.Account, identifier string, config ...ClientConfig
 		}
 	}
 	c := Client{
-		config: _config,
-		account: account,
+		config:    _config,
+		account:   account,
 		publicKey: account.PubKey().EncodePoint(),
 	}
 	c.Address = address.MakeAddressString(c.publicKey, identifier)
@@ -278,6 +281,9 @@ func (c *Client) sendReceipt(prevSignature []byte) error {
 	if err != nil {
 		return err
 	}
+
+	c.Lock()
+	defer c.Unlock()
 	return c.conn.WriteMessage(websocket.BinaryMessage, clientMsgData)
 }
 
@@ -304,11 +310,11 @@ func (c *Client) Send(dests []string, payload []byte, MaxHoldingSeconds ...uint3
 	nonce := randUint32()
 
 	sigChain := pb.SigChain{
-		Nonce:      nonce,
-		DataSize:   uint32(len(payload)),
-		SrcId:      c.addressId,
-		SrcPubkey:  c.publicKey,
-		Elems:      []*pb.SigChainElem{sigChainElem},
+		Nonce:     nonce,
+		DataSize:  uint32(len(payload)),
+		SrcId:     c.addressId,
+		SrcPubkey: c.publicKey,
+		Elems:     []*pb.SigChainElem{sigChainElem},
 	}
 
 	if c.sigChainBlockHash != "" {
@@ -350,7 +356,7 @@ func (c *Client) Send(dests []string, payload []byte, MaxHoldingSeconds ...uint3
 	if err != nil {
 		return err
 	}
-	
+
 	clientMsg := &pb.ClientMessage{
 		MessageType: pb.OUTBOUND_MESSAGE,
 		Message:     outboundMsgData,
@@ -361,6 +367,8 @@ func (c *Client) Send(dests []string, payload []byte, MaxHoldingSeconds ...uint3
 		return err
 	}
 
+	c.Lock()
+	defer c.Unlock()
 	return c.conn.WriteMessage(websocket.BinaryMessage, clientMsgData)
 }
 

